File: rs/src/day8.rs
use std::collections::BTreeSet;
use std::iter::once;

use itertools::Itertools;

fn solve<I>(data: &str, multiples: I) -> usize
where
    I: Clone + Iterator<Item = isize>,
{
    let (mut height, mut width) = (0, 0);
    data.lines()
        .enumerate()
        .flat_map(|(y, line)| {
            height = y + 1;
            width = width.max(line.len());
            line.bytes().enumerate().filter_map(move |(x, b)| match b {
                b'.' => None,
                _ => Some((b, (y, x))),
            })
        })
        .into_grouping_map_by(|(b, _)| *b)
        .fold_with(
            |_, _| vec![],
            |mut acc, _, (_, point)| {
                acc.push(point);
                acc
            },
        )
        .values()
        .flat_map(|points| {
            points.iter().flat_map(|point0 @ (y0, x0)| {
                points
                    .iter()
                    .filter(move |point1| point0 != *point1)
                    .flat_map(|(y1, x1)| {
                        let (dy, dx) = (*y1 as isize - *y0 as isize, *x1 as isize - *x0 as isize);
                        multiples.clone().map_while(move |i| {
                            Some((
                                y1.wrapping_add_signed(i * dy),
                                x1.wrapping_add_signed(i * dx),
                            ))
                            .filter(|(y, x)| *y < height && *x < width)
                        })
                    })
            })
        })
        .collect::<BTreeSet<_>>()
        .len()
}

pub fn part1(data: &str) -> usize {
    solve(data, once(1))
}

pub fn part2(data: &str) -> usize {
    solve(data, 0..)
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        ............
        ........0...
        .....0......
        .......0....
        ....0.......
        ......A.....
        ............
        ............
        ........A...
        .........A..
        ............
        ............
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(14, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(34, part2(EXAMPLE));
    }
}

File: hs/src/Day8.hs
-- |
-- Module:         Day8
-- Description:    <https://adventofcode.com/2024/day/8 Day 8: Resonant Collinearity>
module Day8 (part1, part2) where

import Control.Monad (guard)
import Data.Ix (Ix (inRange))
import Data.Map qualified as Map (elems, fromListWith)
import Data.Set qualified as Set (fromList, singleton, size, toList)
import Data.Text (Text)
import Data.Text qualified as T (length, lines, unpack)

solve :: ((Int, Int) -> (Int, Int) -> [(Int, Int)]) -> Text -> Int
solve extend input =
  Set.size . Set.fromList $ do
    values <- Set.toList <$> Map.elems points
    p0 <- values
    p1 <- values
    guard $ p0 /= p1
    takeWhile (inRange ((0, 0), (height - 1, width - 1))) $ extend p0 p1
  where
    input' = T.lines input
    height = length input'
    width = maximum $ T.length <$> input'
    points =
      Map.fromListWith (<>) $
        [ (c, Set.singleton (y, x))
        | (y, line) <- zip [0 ..] $ T.lines input,
          (x, c) <- zip [0 ..] $ T.unpack line,
          c /= '.'
        ]

part1, part2 :: Text -> Int
part1 = solve $ \(y0, x0) (y1, x1) -> [(2 * y1 - y0, 2 * x1 - x0)]
part2 = solve $ \(y0, x0) (y1, x1) -> zip [y1, 2 * y1 - y0 ..] [x1, 2 * x1 - x0 ..]
